winpr/tools/makecert/makecert.c

```diff
@@ -877,8 +880,7 @@ int makecert_context_output_private_key_file(MAKECERT_CONTEXT* context, char* pa
 	if (!filename)
 		return -1;
 
-	strcpy(filename, context->output_file);
-	strcpy(&filename[length], ".key");
+	sprintf_s(filename, length + 8, "%s.key", context->output_file);
```

Comment:
Why `length+8`?

Yes, filename has been allocated earlier with that size.  But if we have to pass the pointer and the size to all the functions, then it should be reified in a struct {int allocated;char* data} string, and the structure be passed to the string manipulating functions, to avoid any potential inconsistency!

Again why `length+8` ?  We only use literal filename extensions of 4 characters?  Why not length+5?
Or rather, why not just allocate what is needed when it is needed, ie. instead of sprintf use snprintf to compute the required size and allocate it then.  For example something like:
```
#include <stdio.h>
#include <string.h>
#include <stdarg.h>

/* Allocate the required string and format it */ 
int msprintf(char** string, const char * format, ...)
{
	char dummy[2];
	char * result;
	size_t size;
	va_list ap;
	va_start(ap, format);
	/* SUS2 doesn't take size = 0, so we use 1. */ 
	size = vsnprintf(dummy, 1, format, ap);
	va_end(ap);
	result = checked_malloc(size);
	va_start(ap, format);
	size = vsnprintf(result, size, format, ap);
	va_end(ap);
	(*string) = result;
	return size;
}
```

Finally a word about error handling. It is nice to check the result of all allocations. However, I note that:
1- we don't have any memory management that would let us free some memory to continue working properly. (eg. if we had documents, we could close an unused document, free memory, and continue working on the other documents).
2- most of the case, when a memory error occur, we exit with an error code, and ultimately this will close the RDP connection.

Therefore we could as well just exit(1) when an allocation fails. This would simplify.

Reply:
@informatimago Because the `+8` is the allocation size. Might be too big, but didn't want to change it (aka too lazy to check the complete rest of usage, might be nothing, might be something)
As for the structs, have fun refactoring all the code ;)
This is not in scope for 2.0